### DISCUSSION TITLE
Is the whole screen redrawn on every mouse movement? Can anything be done about it?

### DISCUSSION THREAD

**Discussion by janwaltl** [ArthurSonzogni/FTXUI#449] [Q&A]
Hi, I've made a simple example with one dynamic field:
```
#include "ftxui/component/component.hpp"          // for ScreenInteractive, Component
#include "ftxui/component/screen_interactive.hpp" // for ScreenInteractive, Component

int
main() {
    auto screen = ftxui::ScreenInteractive::Fullscreen();
    int value = 0;

    auto status_line = ftxui::Renderer([&] {
        ++value;
        return ftxui::text("status line, counter: " + std::to_string(value)) | ftxui::center;
    });

    auto canvas = ftxui::text("Canvas") | ftxui::center | ftxui::border | ftxui::flex;

    auto layout = ftxui::Container::Vertical(
        {ftxui::Renderer([&] { return canvas; }), ftxui::Renderer([] { return ftxui::separator(); }), status_line});

    screen.Loop(layout);
    return 0;
}
```

When I move mouse to anywhere, the counter goes up so the screen is redrawn. My issue is the continuous mouse movement generates roughly 60% load of one CPU core. That is quite a lot. That does not happen with other TUI applications like `top` even on button presses.

I assume this is caused by running fullscreen and actually redrawing the whole screen even though only a small portion of it changed, is that correct?

Is there any way how to fix this? Maybe somehow signal from a component that it does not need to be redrawn? E.g. `canvas` in my example never changes (if the layout did not change).

Does anyone have any experience what will this do to `ssh` performance on slow connection?

Thanks for help.

**Comment by ArthurSonzogni** [comment]
Event might cause some state to be updated. FTXUI doesn't what users are doing when receiving events. So we draw a new frame after handling events. Please note that it is redrawn only after handling all the pending events. So even if you have 1000 pending events, the screen won't be drawn 1000 times, but only once. This means the drawing rate is "auto-adjusted" to match the performance of the terminal.
In your example, if your terminal was slower, you would get a lower frame count for the same mouse displacement. FTXUI and the terminal can always keep up, no matter the number of events.

If this is an issue to anyone, we could take the result of:
https://github.com/ArthurSonzogni/FTXUI/blob/d9241435ce9fa3d81f0ea5a0d34c3048883629ee/include/ftxui/component/component_base.hpp#L47
It returns a boolean telling if the event was "handled". We could use it to redraw only when users have acknowledged they "handled" the event. Nevertheless, this would break users submitting events to trigger a new frame to be drawn.

**Reply by janwaltl** [reply]
Thanks for the quick answer!

I was intending to use my application primarily as a remote tool, likely even from a serial port. Which puts limitations on how much data can be transferred. If I consider a small 120*60 terminal, then that is 57600 bits for redraw, not counting any escape sequences which there are non-trivial amount I assume. So, we are already talking 1 second redraw time for slower serial connections :( 

Well, I guess I need to properly test it because this library looks very cool and the code is simple - exactly what I am looking for.

Do you think there is an easy way how to disable mouse hover events or limit them explicitly to X/second? Perhaps a custom screen class?

**Reply by ArthurSonzogni** [reply]
> Do you think there is an easy way how to disable mouse hover events or limit them explicitly to X/second? Perhaps a custom screen class?

Maybe I can add some options for you:
```cpp
ScreenInteractive::SetOptions({
     .redraw_after_unhandled_event = false,
     .mouse_support = false,
});
```

Please try developing your application, and if you have performance issues for real, I will add this API.
